### PR TITLE
fix bug with tabview button appearances

### PR DESCRIPTION
**Pull request by patryk-m-malinowski**
This PR fixes an issue with appearances in a qx.ui.tabview.TabView, where if the first tab has its `tabVisibility` set to `excluded`, the first visible tab will not have the state `firstTab`, which will break appearance.

### CODE REVIEW

**Review by cboulanger — APPROVED** [review]
Looks good to me, but I can't test it visually at the moment.

**Review by oetiker — APPROVED** [review]
same here .,.. looks sensible, not testetd

**Review by hkollmann — APPROVED** [review]

**Review by goldim — CHANGES_REQUESTED** [review]
I disagree that it is a **bug**. The first tab is first tab. Why it should move own state to other button otherwise if the first or last button is removed but in case of removal everything is fine. Also could you create an issue next time for discussion please and code example attached to it also would be great bc it is time for other devs to reproduce the issue.

**Comment by johnspackman** [comment]
> Why it should move own state to other button

Because it looks weird from the users point of view - if the first tab's visibility is excluded, then it means that the user's "first" tab does not look like the first when the theme's appearance needs to style that "first" tab differently (eg due to borders etc).  

When a tab widget has a state like "first", this is only something which is passed to the appearance, it seems reasonable that this state is saying which tab "appears first"

**Comment by goldim** [comment]
@johnspackman For me first tab is tab which first in array of tabs and not in visible ones. Maybe users' apps are based on old behavior and I may consider this like broken BC, am I right? I would like hear everyone's opinion on this problem. Maybe to create separate discussion?

**Comment by goldim** [comment]
@p9malino26 In old implementation if I have 3 buttons, removed first and third, central button will have both states. In yours only lastTab.

**Comment by johnspackman** [comment]
by all means create a separate discussion, or we can carry on here; I guess my point is that the state of "first" is only for the appearance, and if the tab is excluded then the appearance will never see a "first" tab (without this PR), in which case what would be the point of a "first" state.

File: source/class/qx/ui/tabview/TabView.js [hunk 1]
```diff
@@ -637,6 +623,32 @@ qx.Class.define("qx.ui.tabview.TabView", {
...
+          if (!firstPage) {
+            firstPage = page;
+          }
+          lastPage = page;
+        }
+        page.removeState("firstTab");
+        page.removeState("lastTab");
+      }
+      firstPage?.addState("firstTab");
+      firstPage?.removeState("lastTab");
```
Comment (goldim) [inline source/class/qx/ui/tabview/TabView.js:648]: You already remove this state in cycle.

File: source/class/qx/ui/tabview/TabView.js [hunk 2]
```diff
@@ -637,6 +623,32 @@ qx.Class.define("qx.ui.tabview.TabView", {
...
+
+      lastPage?.addState("lastTab");
+      lastPage?.removeState("firstTab");
```
Comment (goldim) [inline source/class/qx/ui/tabview/TabView.js:651]: The same

File: source/class/qx/ui/tabview/TabView.js [hunk 3]
```diff
@@ -637,6 +623,32 @@ qx.Class.define("qx.ui.tabview.TabView", {
...
+
+    __onPageChangeTabVisibility() {
+      this.__updateFirstLastTabStates();
+    },
+
+    __updateFirstLastTabStates() {
+      let firstPage = null;
+      let lastPage = null;
+
+      for (let pages = this.getChildren(), i = 0; i < pages.length; i++) {
```
Comment (goldim) [inline source/class/qx/ui/tabview/TabView.js:636]: Could you move `let pages = this.getChildren()` out of cycle on separate line, please? Several instructions on one line - looks like a bad smell.

**Review by goldim — CHANGES_REQUESTED** [review]

File: source/class/qx/ui/tabview/TabView.js (hunk 3 above)
Comment (johnspackman) [reply source/class/qx/ui/tabview/TabView.js:636]: we use that pattern in a lot of places - I'm counting 290 `for` loops with multiple variables declared, one for the array and one for the index.  My regex is `for \((var|let) [^;]+,`

Comment (goldim) [reply source/class/qx/ui/tabview/TabView.js:636]: Ok, I won't insist. Just noticed. My view on good code is obviously different. May fix it, it is up to you.

**Comment by goldim** [comment]
@johnspackman I can see that it is a bug actually. Classic Theme:
image

**Comment by goldim** [comment]
@p9malino26 just fix what I suggested above and I gonna approve it.

**Review by goldim — APPROVED** [review]

**Review by derrell — APPROVED** [review]

**Pull request merged**